### ISSUE TITLE
webpack 5.0.0-beta.30 Module not found: Error: Can't resolve '../../core-js/object/define-property'

### ISSUE THREAD

**Issue by noe132** [webpack/webpack#11467]
# Bug report

**What is the current behavior?**

install `@babel/runtime-corejs3`
for file `index.js`
```js
import '@babel/runtime-corejs3/helpers/esm/defineProperty'
```
and run `webpack ./index.js`

```
ERROR in ./node_modules/@babel/runtime-corejs3/helpers/esm/defineProperty.js 1:0-74
Module not found: Error: Can't resolve '../../core-js/object/define-property' in 'xxx\node_modules\@babel\runtime-corejs3\helpers\esm'
 @ ./index.js 1:0-58
```

`beta.29` is working properly on this.

but if we manually edit line 1 in file `./node_modules/@babel/runtime-corejs3/helpers/esm/defineProperty.js`
from
```js
import _Object$defineProperty from "../../core-js/object/define-property";
```
to
```js
import _Object$defineProperty from "../../core-js/object/define-property.js";
```
and this issue goes away.

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**

**Other relevant information:**
webpack version:  5.0.0-beta.30
Node.js version: 12.16.1
Operating System: win10 2004
Additional tools:
at the time I'm using `@babel/runtime-corejs3@^7.11.2`

**Comment by EvHaus** [comment]
I'm having a very similar issue after upgrading from beta.29 to beta.30, but with this file:

```
ERROR in ./node_modules/graphql/language/parser.mjs 1:0-41
Module not found: Error: Can't resolve '../jsutils/inspect' in '/Users/evhaus/Git/zenhub/client/node_modules/graphql/language'
 @ ./node_modules/graphql-tag/src/index.js 1:13-47
 @ ./shared/services/GitHubGraphQL/fetchIssueDataAsREST.js 23:0-30 24:12-15
 @ ./shared/services/GitHubGraphQLAPI.js 25:0-79 130:9-29
 @ ./shared/services/Issue.js 6:0-59 67:9-37 73:9-41
 @ ./shared/store/actions/types/boardActions.js 25:0-47 194:77-102 197:75-93 202:81-102 251:89-118 252:86-115
 @ ./shared/store/ducks/workspaceForm.js 70:0-83 1115:17-30 1974:15-31
 @ ./webapp/src/pages/RedirectToWorkspaceOrCreate.jsx 34:0-76 230:15-56
 @ ./webapp/src/Root.tsx 39:0-78 227:15-42
 @ ./webapp/src/index.jsx 23:0-26 68:7-11
```

I'm guessing the cache isn't being reset properly? What's the right way to clear it?

**Comment by EvHaus** [comment]
I'm having a very similar problem after upgrading from beta.29 to beta.30 but with this different import:

```
ERROR in ./node_modules/graphql/language/parser.mjs 1:0-41
Module not found: Error: Can't resolve '../jsutils/inspect' in '/Users/my_project/node_modules/graphql/language'
 @ ./node_modules/graphql-tag/src/index.js 1:13-47
 @ ./webapp/src/index.jsx 23:0-26 68:7-11
```

Clearing the local filesystem cache (`rm -rf node_modules/.cache`) didn't help. But rolling back to beta.29 worked for me as well.

**Comment by duan602728596** [comment]
I also encountered the same problem, which seems to be caused by enabling mjs experiments, and I did not find a way to close mjs experiments.

**Comment by sokra** [comment]
These are all the same issues. In .mjs files or .js files with a package.json with `"type": "module"` imports need to be fully specified. This means you need to have an extension.

That how ESM in node.js is defined, and you are opting in into this behavior by using the .mjs extension resp the field in the package.json.

We didn't enforce this behavior in previous versions while this was still experimental in node.js, but it left this status, so I feel ok with enforcing this and pointing out these incorrect packages.

You could disable that with `{ test: /\.m?js/, type: "javascript/auto" }` in `module.rules`.

**Comment by ocdi** [comment]
@sokra I have extremely similar symptoms, even with the above `module.rules` suggestion. Rolling back to `beta.29` compiles successfully

```
ERROR in ./node_modules/@babel/runtime/helpers/esm/toConsumableArray.js 3:0-70
Module not found: Error: Can't resolve './unsupportedIterableToArray' in 'C:\s\node_modules\@babel\runtime\helpers\esm'
```

I tried various positions of adding the rule, setting type in other potential rules involving .js/typescript that might affect it, but none solved this module not found issue.

I suspect babel need to fix their code to support this correctly, otherwise there will be many people having issues with this on webpack5.

For now I am sticking to `beta.29`.

**Comment by duan602728596** [comment]
After configuring like this, there will still be some modules Can't resolve.

```javascript
const path = require('path');

module.exports = {
  mode: 'development',
  entry: { index: path.join(__dirname, 'src/index.js') },
  devtool: 'source-map',
  module: {
    rules: [
      { test: /\.m?js/, type: "javascript/auto" },
      {
        test: /.*\.js$/,
        type: 'javascript/auto',
        exclude: /node_modules/,
        use: [
          {
            loader: 'babel-loader',
            options: {
              presets: [
                [
                  '@babel/preset-env',
                  {
                    targets: {
                      browsers: ['IE 11']
                    },
                    debug: false,
                    modules: false,
                    useBuiltIns: 'usage',
                    bugfixes: true,
                    corejs: 3
                  }
                ]
              ],
              plugins: [
                [
                  '@babel/plugin-transform-runtime',
                  {
                    corejs: { version: 3, proposals: true },
                    helpers: true,
                    regenerator: true,
                    useESModules: true
                  }
                ]
              ]
            }
          }
        ]
      }
    ]
  }
};
```

```javascript
import _asyncToGenerator from '@babel/runtime-corejs3/helpers/esm/asyncToGenerator.js';

function main() {
  console.log(_asyncToGenerator);
}

main();
```

**Comment by sokra** [comment]
btw this is not an webpack-only error. If you run `import("@babel/runtime-corejs3/helpers/esm/defineProperty.js")` in node.js you get an error too:

```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '...\node_modules\@babel\runtime-corejs3\core-js\object\define-property' imported from ...\node_modules\@babel\runtime-corejs3\helpers\esm\defineProperty.js
```

---

> You could disable that with `{ test: /\.m?js/, type: "javascript/auto" }` in `module.rules`.

Ok I tried that by myself and it doesn't work correctly. Instead you have to do that to change the resolving for these modules:

```
{
  test: /\.m?js/,
  resolve: {
    fullySpecified: false
  }
}
```

`type: "javascript/auto"` might be added to fully disable the different behavior for .mjs files, but that doesn't affect the resolving. It only affects e. g. how interop with non-ESM modules is handled.

**Comment by evilebottnawi** [comment]
@sokra maybe we can put this option to https://webpack.js.org/configuration/resolve/ (by default `true`), I'm afraid of too many projects use wrong paths in `import`, using:
```
{
  test: /\.m?js/,
  resolve: {
    fullySpecified: false
  }
}
```

with babel/ts/other loaders will increase the size an complex of configurations

**Comment by evilebottnawi** [comment]
Also I think we need docs this

/cc @chenxsan

**Comment by vankop** [comment]
@evilebottnawi 

> @sokra maybe we can put this option to https://webpack.js.org/configuration/resolve/ (by default true), I'm afraid of too many projects use wrong paths in import, using:

thats how Node.js resolves.. Maybe we need to point in docs workaround..

**Comment by evilebottnawi** [comment]
@vankop yes, I known, in theory we can improve error messages out of box, like:
```
ERROR in ./node_modules/@babel/runtime-corejs3/helpers/esm/defineProperty.js 1:0-74
Module not found: Error: Can't resolve '../../core-js/object/define-property' in 'xxx\node_modules\@babel\runtime-corejs3\helpers\esm'
 @ ./index.js 1:0-58

You're using ES module syntax, maybe you meant:
import '@babel/runtime-corejs3/helpers/esm/defineProperty.js'
or
import '@babel/runtime-corejs3/helpers/esm/defineProperty.mjs'
```

Because now you might think something wrong with resolving

**Comment by JamesAlen9352** [comment]
![](https://media2.giphy.com/media/l46Czzp0KEHSO7OdG/giphy.gif)

**Comment by JamesAlen9352** [comment]
![](https://media3.giphy.com/media/3ktLJJs5fYT6NP005b/giphy-downsized-medium.gif)

**Comment by JamesAlen9352** [comment]
![](https://media1.giphy.com/media/l0K4cvnVbkOYhvyyA/giphy.gif)

**Comment by otakustay** [comment]
I also encounter similar issue but the message is like:

```
error  in ./node_modules/@babel/runtime/helpers/esm/toConsumableArray.js

Module not found: Error: Can't resolve './iterableToArray' in '/xxx/node_modules/@babel/runtime/helpers/esm'
```

To what I know this is an error when `@babel/runtime` is importing its internal module (both inside `helpers/esm` folder), es module resolution of node should not forbid internal imports, it only affects how other modules import it from external.

**Comment by sokra** [comment]
It's not forbidden. The import request need to be fully specified as `'./iterableToArray.js'`.

**Comment by wojtekmaj** [comment]
Happening to me as well. Upgrading Webpack from beta.29 to beta.30, not doing literally anything else, breaks the build:

```
ModuleNotFoundError: Module not found: Error: Can't resolve './assertThisInitialized' in '/node_modules/@babel/runtime/helpers/esm'
    at /node_modules/webpack/lib/Compilation.js:1470:28
    at /node_modules/webpack/lib/NormalModuleFactory.js:647:13
    at eval (eval at create (/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:10:1)
    at /node_modules/webpack/lib/NormalModuleFactory.js:233:22
    at eval (eval at create (/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:9:1)
    at /node_modules/webpack/lib/NormalModuleFactory.js:357:22
    at /node_modules/webpack/lib/NormalModuleFactory.js:116:11
    at /node_modules/webpack/lib/NormalModuleFactory.js:576:24
    at finishWithoutResolve (/node_modules/enhanced-resolve/lib/Resolver.js:284:11)
    at /node_modules/enhanced-resolve/lib/Resolver.js:350:15
    at /node_modules/enhanced-resolve/lib/Resolver.js:398:5
    at eval (eval at create (/node_modules/enhanced-resolve/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:15:1)
    at /node_modules/enhanced-resolve/lib/Resolver.js:398:5
    at eval (eval at create (/node_modules/enhanced-resolve/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:27:1)
    at /node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:87:43
    at /node_modules/enhanced-resolve/lib/Resolver.js:398:5
    at eval (eval at create (/node_modules/enhanced-resolve/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:16:1)
    at /node_modules/enhanced-resolve/lib/Resolver.js:398:5
    at eval (eval at create (/node_modules/enhanced-resolve/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:27:1)
    at /node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:87:43
    at /node_modules/enhanced-resolve/lib/Resolver.js:398:5
    at eval (eval at create (/node_modules/enhanced-resolve/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:16:1)
    at /node_modules/enhanced-resolve/lib/ConditionalPlugin.js:39:48
    at _next0 (eval at create (/node_modules/enhanced-resolve/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:8:1)
    at eval (eval at create (/node_modules/enhanced-resolve/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:30:1)
    at /node_modules/enhanced-resolve/lib/Resolver.js:398:5
    at eval (eval at create (/node_modules/enhanced-resolve/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:57:1)
    at /node_modules/enhanced-resolve/lib/ConditionalPlugin.js:52:42
    at /node_modules/enhanced-resolve/lib/Resolver.js:398:5
    at eval (eval at create (/node_modules/enhanced-resolve/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:16:1)
resolve './assertThisInitialized' in '/node_modules/@babel/runtime/helpers/esm'
  using description file: /node_modules/@babel/runtime/helpers/esm/package.json (relative path: .)
    Field 'browser' doesn't contain a valid alias configuration
    using description file: /node_modules/@babel/runtime/helpers/esm/package.json (relative path: ./assertThisInitialized)
      Field 'browser' doesn't contain a valid alias configuration
      /node_modules/@babel/runtime/helpers/esm/assertThisInitialized doesn't exist
```

**Comment by nickbouton** [comment]
Also seeing this issue with beta 30->32 as well, 29 still works ok. All of the errors we're getting are coming from Material UI via @babel/runtime.

**Comment by evilebottnawi** [comment]
There is no problems on webpack side

**Comment by EvHaus** [comment]
I'm not having luck with the `fullySpecified` solution. When I add that to my module rules and try the latest `rc.0` build I end up with this failure:

```
/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:362
			newStack = new Set(resolveContext.stack);
			           ^

RangeError: Maximum call stack size exceeded
    at new Set (<anonymous>)
    at Resolver.doResolve (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:362:15)
    at resolver.getHook.tapAsync (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/ConditionalPlugin.js:42:14)
    at Hook.eval [as callAsync] (eval at create (/myproject/client/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:37:1)
    at Resolver.doResolve (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:394:16)
    at resolver.getHook.tapAsync (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/NextPlugin.js:30:14)
    at Hook.eval [as callAsync] (eval at create (/myproject/client/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:7:1)
    at Resolver.doResolve (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/Resolver.js:394:16)
    at DescriptionFileUtils.loadDescriptionFile (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:74:17)
    at forEachBail (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFileUtils.js:118:13)
```

And sometimes this one:

```
/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFileUtils.js:42
	(function findDescriptionFile() {
	                             ^

RangeError: Maximum call stack size exceeded
    at findDescriptionFile (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFileUtils.js:42:31)
    at forEachBail (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFileUtils.js:125:14)
    at iterator (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/forEachBail.js:16:12)
    at resolver.fileSystem.readJson (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFileUtils.js:58:16)
    at CacheBackend.provide (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:174:31)
    at forEachBail (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFileUtils.js:52:26)
    at next (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/forEachBail.js:14:3)
    at forEachBail (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/forEachBail.js:24:9)
    at findDescriptionFile (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFileUtils.js:47:3)
    at forEachBail (/myproject/client/node_modules/webpack/node_modules/enhanced-resolve/lib/DescriptionFileUtils.js:125:14)
```

The same issues don't happen on beta.29. Is anyone else seeing similar problems?

**Comment by evilebottnawi** [comment]
@EvHaus Can you create reproducible test repo?

**Comment by EvHaus** [comment]
Logged as a new issue with reproducible test repo here: https://github.com/webpack/webpack/issues/11518 I think it's unrelated to this issue actually.

**Comment by X-neuron** [comment]
I'm having the similar problem:

webpack version: 5.0.0-rc.0
node version:14.11.0

.babelrc.js
```
"plugins": [
    [
      "@babel/plugin-transform-runtime",
      {
        "absoluteRuntime": false,
        "corejs": 3,
        "helpers": true,
        "regenerator": true,
        "useESModules": true
      }
    ],
  ]

```
```
ERROR in ./node_modules/@babel/runtime-corejs3/helpers/esm/jsx.js 1:0-51
Module not found: Error: Can't resolve '../../core-js/symbol/for' in 'E:\project\antdFront2\node_modules\@babel\runtime-corejs3\helpers\esm'
Did you mean 'for.js'?
BREAKING CHANGE: The request '../../core-js/symbol/for' failed to resolve only because it was resolved as fully specified
(probably because the origin is a '*.mjs' file or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.
```

**Comment by mkermani144** [comment]
I have the same issue as @wojtekmaj. `{ fullySpecified: false }` doesn't work for me.

**Comment by SokichiFujita** [comment]
I've been investigating and trying workarounds. However the error still exists with stable `v5.0.0`. `{ fullySpecified: false }` or another workarounds which are proposed in related issue does not work well now.

**Comment by evilebottnawi** [comment]
@SokichiFujita it means you have other broken packages please provide screenshots or the error message

**Comment by SokichiFujita** [comment]
@evilebottnawi Thanks. Does it mean webpack 5 has no problem but material-ui or babel (plugins) has problem?

<img width="1382" alt="webpackv5" src="https://user-images.githubusercontent.com/15647722/95683032-acf41c80-0c23-11eb-9e04-10cfb3520314.png">

Error messages:

```
ERROR in ./node_modules/@babel/runtime/helpers/esm/objectWithoutProperties.js 1:0-74
Module not found: Error: Can't resolve './objectWithoutPropertiesLoose' in '/Users/s/mydev/react/starter-react-flux/__test/node_modules/@babel/runtime/helpers/esm'
Did you mean 'objectWithoutPropertiesLoose.js'?
BREAKING CHANGE: The request './objectWithoutPropertiesLoose' failed to resolve only because it was resolved as fully specified
(probably because the origin is a '*.mjs' file or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.
 @ ./node_modules/@material-ui/core/esm/styles/transitions.js 1:0-90 57:16-40
 @ ./node_modules/@material-ui/core/esm/styles/index.js 9:0-30 9:0-30
 @ ./app/App.js 4:0-76 7:15-29 18:41-57
 :

same messages 
```

**Comment by evilebottnawi** [comment]
Yes, it is bug on Babel side, they will do release on the next week

**Comment by SokichiFujita** [comment]
@evilebottnawi I got it. Maybe this issue https://github.com/babel/babel/issues/12058 . The next babel 7.12 may solve these  issue.

**Comment by gtwilliams03** [comment]
> @sokra maybe we can put this option to https://webpack.js.org/configuration/resolve/ (by default `true`), I'm afraid of too many projects use wrong paths in `import`, using:
> 
> ```
> {
>   test: /\.m?js/,
>   resolve: {
>     fullySpecified: false
>   }
> }
> ```
> 
> with babel/ts/other loaders will increase the size an complex of configurations

This worked for me - but I needed to add a dollar sign to the end of the regular expression (`/\.m?js$/`) or webpack 5 started failing on JSON files (it matched .json as well as .mjs and .js).

**Comment by ScriptedAlchemy** [comment]
@sokra it might be worth pinning this issue. I’m getting a lot of traffic around mjs resolution, everyone needs the same thing

**Comment by evilebottnawi** [comment]
@ScriptedAlchemy Already pinned https://github.com/webpack/webpack :smile:

**Comment by doberkofler** [comment]
This problem was resolved by using this role but only after upgrading to webpack 5.1.0
```
{
  test: /\.m?js/,
  resolve: {
    fullySpecified: false
  }
}
```

**Comment by bertho-zero** [comment]
@SokichiFujita You may have added `fullySpecified` in the wrong place, I made this error, it is not the general resolve but that of a rule:

```js
    module: {
      rules: [
        {
          test: /\.(js|mjs|jsx)$/,
          enforce: 'pre',
          loader: require.resolve('source-map-loader'),
          resolve: {
            fullySpecified: false
          }
        },
        // ...
      ]
    }
```

**Comment by joebnb** [comment]
similar configuration still not working

**Comment by ScriptedAlchemy** [comment]
You need extensions as well most likely. Or to include node modules

**Comment by doberkofler** [comment]
I really just used this role in 5.1.0 and it worked
```
{
  test: /\.m?js$/,
  resolve: {
    fullySpecified: false
  }
}
```

**Comment by kagawagao** [comment]
> similar configuration still not working with webpack 5.1.0
> 
> ```js
> {
> test: /\.(mj|j|t)sx?$/,
> resolve: {
>     fullySpecified: false
>   },
>   type: "javascript/auto",
>   exclude: /node_modules/,
>   use: {
>     loader: "babel-loader",
>     options: {
>       cwd: BABEL_PATH,
>       babelrcRoots: BABEL_PATH,
>       cacheDirectory: true,
>       babelrc: false,
> ... ...
> }
> ```

make `fullySpecified` to works on all files, not only source files, but also `node_modules`, like:
```js
{
        test: /\.m?js/,
        resolve: {
          fullySpecified: false,
        },
      },
      {
        test: /\.(j|t)sx?$/,
        include: [srcDir],
        exclude: /node_modules/,
        use: [
          {
            loader: 'babel-loader',
            options: {
              cacheDirectory: true,
            },
          },
        ],
      }

```

**Comment by maurobalbi** [comment]
> > similar configuration still not working with webpack 5.1.0
> > ```js
> > {
> > test: /\.(mj|j|t)sx?$/,
> > resolve: {
> >     fullySpecified: false
> >   },
> >   type: "javascript/auto",
> >   exclude: /node_modules/,
> >   use: {
> >     loader: "babel-loader",
> >     options: {
> >       cwd: BABEL_PATH,
> >       babelrcRoots: BABEL_PATH,
> >       cacheDirectory: true,
> >       babelrc: false,
> > ... ...
> > }
> > ```
> 
> make `fullySpecified` to works on all files, not only source files, but also `node_modules`, like:
> 
> ```js
> {
>         test: /\.m?js/,
>         resolve: {
>           fullySpecified: false,
>         },
>       },
>       {
>         test: /\.(j|t)sx?$/,
>         include: [srcDir],
>         exclude: /node_modules/,
>         use: [
>           {
>             loader: 'babel-loader',
>             options: {
>               cacheDirectory: true,
>             },
>           },
>         ],
>       }
> ```

This solution fixed the problem for me. Thanks alot.

**Comment by evilebottnawi** [comment]
Fixed https://babeljs.io/blog/2020/10/15/7.12.0, please leave feedback

**Comment by llooz** [comment]
@evilebottnawi just upgraded to babel 7.12 and removed the `fullySpecified` rule. Works as intended 👍

**Comment by sokra** [comment]
Awesome

**Comment by doberkofler** [comment]
@llooz I actually do not use Babel in my build but still had the same problem.
Do I need to keep the
```
{
  test: /\.m?js$/,
  resolve: {
    fullySpecified: false
  }
}
```
rule until all my dependencies upgrade Babel?

**Comment by llooz** [comment]
@doberkofler for me it was the combination of webpack 5 + babel 7.11 + preset-env + core-js@3 that surfaced the issue.
Once they've fixed the ESM handling on babel 7.12 the issue disappeared.
Unfortunately I can't think of what can it be the issue in your project.

**Comment by anajavi** [comment]
> @llooz I actually do not use Babel in my build but still had the same problem.
> Do I need to keep the
> 
> ```
> {
>   test: /\.m?js$/,
>   resolve: {
>     fullySpecified: false
>   }
> }
> ```
> 
> rule until all my dependencies upgrade Babel?

You might need to refresh your npm/yarn lockfile.

Check that your currently installed runtime is at least 7.12.0  with ```npm ls @babel/runtime```

**Comment by ScriptedAlchemy** [comment]
This error happens most commonly with Babel. Which is now fixed. But other packages that have not fixed this issue will require that mjs resolution or you want to contact the project author and ask them to update

**Comment by doberkofler** [comment]
What really confuses me, is the fact that I actually do not actually have any rules that would use Babel.
Does this problem also arise when webpack just uses packages that have been build with Babel?

**Comment by bertho-zero** [comment]
@doberkofler For my part I have dependencies which use `@babel/runtime-corejs2`, I had to remove the entry from `yarn.lock` then run `yarn` to install the latest version.

**Comment by otakustay** [comment]
> What really confuses me, is the fact that I actually do not actually have any rules that would use Babel.

Some third party packages may contain code compiled via babel, and those file can have an import statement to `@babel/runtime`

**Comment by bertho-zero** [comment]
This is what I describe just above, these dependencies are normally defined with a `^` and the lockfile (yarn.lock or package-lock.json) blocks them, you have to remove the entry from the lockfile and then restart an installation to that these update.

**Comment by evilebottnawi** [comment]
If somebody needs help please provide error message

**Comment by SokichiFujita** [comment]
@bertho-zero Thank you. Regarding previous babel 7.11.6 and webpack v5.0.0, I also had been tried the following configuration. But it was failed.

```
  module: {
    rules: [
      {
        test: /.(js|jsx|mjs)$/,
        exclude: [path.resolve(__dirname, "node_modules")],
        loader: "babel-loader",
        resolve: {
          fullySpecified: false
        }
      },
    ],
  },
 ```

**Comment by SokichiFujita** [comment]
With Babel 7.12.0 and Webpack v5.1.0, completely works well. Thank you!

**Comment by SokichiFujita** [comment]
babel v7.12.0 and webpack v5.0.0 also work well without workarounds. It was issue of old babel.

**Comment by doberkofler** [comment]
I don't use babel in any of my rules nor is it a dependency in my package.json file.
I did remove my package-lock.json file to update all dependencies.
I am using webpack 5.1.0 but still get a large number of errors like the following when not using this rule
```js
{
	test: /\.m?js$/,
	resolve: {
		fullySpecified: false
	}
}
```

**errors**
```
assets by status 100 MiB [cached] 509 assets
orphan modules 1.53 MiB [orphan] 1026 modules
runtime modules 246 KiB 435 modules
modules by path ./ 14.2 MiB (javascript) 591 KiB (css/mini-extract)
  javascript modules 14.2 MiB 1259 modules
  css modules 591 KiB
    modules by path ./src/ 342 KiB 61 modules
    modules by path ./vendor/ 167 KiB 8 modules
    modules by path ./node_modules/ 81.6 KiB 8 modules
  json modules 62.9 KiB
    modules by path ./node_modules/react-html-parser/node_modules/entities/maps/*.json 30 KiB 4 modules
    modules by path ./node_modules/entities/lib/maps/*.json 30 KiB 4 modules
    ./node_modules/dom-serializer/foreignNames.json 2.79 KiB [built] [code generated]
readable-stream (ignored) 15 bytes [built] [code generated]

ERROR in ./node_modules/@fullcalendar/common/main.js 9:0-86
Module not found: Error: Can't resolve './vdom' in '/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/@fullcalendar/common'
Did you mean 'vdom.js'?
BREAKING CHANGE: The request './vdom' failed to resolve only because it was resolved as fully specified
(probably because the origin is a '*.mjs' file or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.
 @ ./node_modules/@fullcalendar/core/main.js 8:0-212 9:0-37 9:0-37 36:16-22 36:23-36 36:37-49 39:28-41 39:42-77 40:24-37 40:38-53 45:16-22 49:12-22 52:33-46 53:12-31 88:8-18 105:13-26 119:8-22 122:2-13
 @ ./src/modules/las_mod_dashboard/widget_calendar/index.ts 12:0-46 28:25-33
 @ ./src/modules/las_mod_dashboard/index.ts 29:0-67 63:16-36

ERROR in ./node_modules/@fullcalendar/core/main.js 6:0-16
Module not found: Error: Can't resolve './vdom' in '/Users/doberkofler/MyDev/ljs_app/trunk/periscope/node_modules/@fullcalendar/core'
Did you mean 'vdom.js'?
BREAKING CHANGE: The request './vdom' failed to resolve only because it was resolved as fully specified
(probably because the origin is a '*.mjs' file or a '*.js' file where the package.json c...ontains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.
 @ ./src/modules/las_mod_dashboard/widget_calendar/index.ts 12:0-46 28:25-33
 @ ./src/modules/las_mod_dashboard/index.ts 29:0-67 63:16-36
...
```

**Comment by evilebottnawi** [comment]
@doberkofler please read the error message, problem in `fullcalendar`

https://github.com/fullcalendar/fullcalendar/blob/master/packages/core/package.json#L12

But

https://github.com/fullcalendar/fullcalendar/blob/master/packages/core/src/main.ts#L2 :

```js
import './vdom'
export { Calendar } from './Calendar'
export * from '@fullcalendar/common'
```

`import './vdom'` is invalid and even works on node, it should be `import './vdom.js'`

**Comment by doberkofler** [comment]
@evilebottnawi So to summarize: this problem can also appear when not explicitly using Babel in 3rd party dependencies and should be reported to the individual maintainers to slowly fix this issue.

**Comment by evilebottnawi** [comment]
@doberkofler yes, because it is invalid, this code **will not work** on node or browser, so webpack starts to use strict rules for this, even it is not webpack problem, try to use `node@14` and run this code, you will get the same problem

**Comment by chooban** [comment]
I have what I think might be a similar problem using `apollo-upload-client`, which uses `extract-files`:

```
ERROR in ../../../node_modules/apollo-upload-client/public/createUploadLink.js 18:19-63
Module not found: Error: Can't resolve 'extract-files/public/extractFiles' in '/Users/ross/Projects/project/node_modules/apollo-upload-client/public'
 @ ./src/app/state/client.ts 844:0-79 931:42-58
 @ ./src/app/views/app/AppContainer.tsx 283:0-40 335:14-20
 @ ./src/app.js 556:0-56 641:9-21
```

The author is adamant that it's [not a problem with their library](https://github.com/jaydenseric/extract-files/issues/15).

I have the `fullySpecificed: false` rule in place, but it's not making a difference. Since the issue lies in a third party library and not our source code, do I need to modify the additional rule to specifically target that library?

**Comment by evilebottnawi** [comment]
@chooban this problem not related to original problem, already reported https://github.com/webpack/enhanced-resolve/issues/256

**Comment by craigmiller160** [comment]
Hey, I just wanted to chime in with something that might help folks. I just upgraded a home-grown config library to Webpack 5, bumped all my dependencies, and have been integrating this into my various projects. A JavaScript project using @material-ui exploded with these errors when I did this.

The resolve/fullySpecified fix mentioned above did nothing for me. I was stuck here for a bit, but then I found something that did work.

I'm using yarn, and there's the "resolutions" option you can add to the package.json. @material-ui was pulling in @babel/runtime 7.11.1, and I wanted to make sure it was on the latest 7.12.5. So I added this resolution to my package.json:

```
"resolutions": {
    "@babel/runtime": "^7.12.5"
  },
```

And viola, everything is fixed now. The fullySpecified option isn't even needed anymore. The app appears to be working fine, although more testing is definitely required.

**Comment by trusktr** [comment]
> You could disable that with `{ test: /\.m?js/, type: "javascript/auto" }` in `module.rules`.

I tried that, but I get the same errors.

My config has it:

```js
const path = require("path");

module.exports = {
    entry: "./dist/main.js",
    output: {
        path: path.resolve(__dirname, "dist"),
        filename: "bundle.js",
    },
    externals: ["fs"],
    module: {
        rules: [{ test: /\.m?js$/, type: "javascript/auto" }],
    },
};
```

and I still get

```
BREAKING CHANGE: The request './dom' failed to resolve only because it was resolved as fully specified
(probably because the origin is a '*.mjs' file or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.
```

What's missing?

**Comment by trusktr** [comment]
For the time being, downgrading to Webpack 4 was easier.

**Comment by alexander-akait** [comment]
> I'm really looking forward to when the ecosystem isn't so fragmented around native ESM. That'll be very nice!

There is no problems on webpack side, please read the error message, problem in `@lume/element` package, they used ES module syntax, but don't use `.js`/`.mjs` extensions in `import`, try to run this code in node or browsers without webpack, you will get the problems

**Comment by 4074** [comment]
@trusktr Try to install the latest `@babel/runtime` by `npm i @babel/runtime -D`, it works for me.

**Comment by thw0rted** [comment]
I'm pretty surprised that we got this many comments in without anybody linking to https://github.com/microsoft/TypeScript/issues/16577 -- I'd bet dollars to doughnuts that all the libraries that people are having trouble with were compiled from Typescript, which 1) encourages users to write extensionless relative import specifiers, then 2) does not give any way to convert those to valid (extension-included) specifiers.

(I'm also having this problem with a compiled-from-TS library...)

**Comment by zachsa** [comment]
Having tried quite a few things to debug this, I found eventually that deleting lockfile and node_modules and reinstalling resulted in a working build

**Comment by thw0rted** [comment]
Zach, that very likely means that one of your deps had a bad import specifier which they fixed in a later (but "compatible") release.

**Comment by trusktr** [comment]
I believe that `fullySpecified: true` is a very good default option, because it encourages everyone to work towards the future of ES Modules and to stop supporting anything else that is not spec-compliant. Library authors will have incentive to fix their libs so that they work in Webpack with defaults that enforce ES Module standards. This will make the ecosystem a lot better.

Imagine if `fullySpecified: false`, we'd have a _whole new generation_ of Webpack where people are writing code that isn't ES Module compliant, and therefore encouraging further fragmentation between real ES Module code and _ES-Module-like_ code.

**Comment by valkn0t** [comment]
Ran into this issue with Babel 7.11.~ after updating to Webpack 5.
Enumerating the error messages here so other people looking for help will find the fix here. 
```
Can't resolve './objectWithoutPropertiesLoose'
Can't resolve './arrayWithHoles'
Can't resolve './iterableToArrayLimit'
Can't resolve './unsupportedIterableToArray'
Can't resolve './nonIterableRest'
Can't resolve './arrayWithoutHoles'
Can't resolve './iterableToArray'
Can't resolve './unsupportedIterableToArray'
Can't resolve './nonIterableSpread'

BREAKING CHANGE: The request './<module>' failed to resolve only because it was resolved as fully specified
```
## Fix (as mentioned above):

Add the following rule to your `webpack.config.js`: 

```
--snip--
 module: {
    rules: [
        {
            test: /\.m?js/,
            resolve: {
                fullySpecified: false
            }
        },
--snip--
```

**Comment by baconcheese113** [comment]
`babel-plugin-proposal-private-methods` from [babel/core@7.14.0](https://github.com/babel/babel/releases/tag/v7.14.0) worked as well and doesn't require the extra webpack rule for node_modules
```
{
  "env": {
    "development": {
      "plugins": [
        [
          "@babel/plugin-proposal-private-methods",
          {
            "loose": true
          }
        ]
      ]
    },
  }
}
```

**Comment by GeekyMonkey** [comment]
If you're using chainWebpack, then it's like this

chain.module.rule("mjs-rule").test(/\.m?js/).resolve.set(fullySpecified, false);

**Comment by MiguelNiblock** [comment]
> btw this is not an webpack-only error. If you run `import("@babel/runtime-corejs3/helpers/esm/defineProperty.js")` in node.js you get an error too:
> 
> ```
> Error [ERR_MODULE_NOT_FOUND]: Cannot find module '...\node_modules\@babel\runtime-corejs3\core-js\object\define-property' imported from ...\node_modules\@babel\runtime-corejs3\helpers\esm\defineProperty.js
> ```
> 
> > You could disable that with `{ test: /\.m?js/, type: "javascript/auto" }` in `module.rules`.
> 
> Ok I tried that by myself and it doesn't work correctly. Instead you have to do that to change the resolving for these modules:
> 
> ```
> {
>   test: /\.m?js/,
>   resolve: {
>     fullySpecified: false
>   }
> }
> ```
> 
> `type: "javascript/auto"` might be added to fully disable the different behavior for .mjs files, but that doesn't affect the resolving. It only affects e. g. how interop with non-ESM modules is handled.

This fixed it for me

**Comment by mudassir-jmi** [comment]
(webpack-internal:///./node_modules/next/dist/pages/_document.js:16:1)

this error show in next js project

**Comment by tayyab-ash** [comment]
Can anyone solve this please... i am getting it in the browser console :)

`Uncaught TypeError: (0 , codecs_data_structures_1.getBytesCodec) is not a function
    at ./node_modules/@solana/spl-token-metadata/lib/cjs/state.js (state.ts:7:1)
    at options.factory (react refresh:6:1)
    at __webpack_require__ (bootstrap:22:1)
    at fn (hot module replacement:61:1)
    at ./node_modules/@solana/spl-token-metadata/lib/cjs/index.js (index.ts:4:1)
    at options.factory (react refresh:6:1)
    at __webpack_require__ (bootstrap:22:1)
    at fn (hot module replacement:61:1)
    at ./node_modules/@metaplex-foundation/mpl-token-metadata/node_modules/@solana/spl-token/lib/cjs/extensions/tokenMetadata/actions.js (actions.ts:4:1)
    at options.factory (react refresh:6:1)`
